src/librustc_typeck/check/_match.rs

```diff
@@ -659,6 +659,12 @@ pub fn check_pat_enum<'a, 'tcx>(pcx: &pat_ctxt<'a, 'tcx>,
     let report_bad_struct_kind = |is_warning| {
         bad_struct_kind_err(tcx.sess, pat.span, path, is_warning);
         if is_warning {
+            // Boo! Too painful to attach this to the actual warning,
```

Comment:
Hmm, I think we should modify `bad_struct_kind_err` to look at `is_warning` and add a lint in the case of `is_warning == true`. This will also require adding a new lint corresponding to this future incompatibility. That said, I was going to do a quick PR to improve that tooling so as to address https://github.com/rust-lang/rust/issues/30746, so maybe I could just roll that change into that.